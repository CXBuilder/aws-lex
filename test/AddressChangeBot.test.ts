import * as cdk from 'aws-cdk-lib';
import { Template, Match } from 'aws-cdk-lib/assertions';
import { AddressChangeBot } from './AddressChangeBot';

describe('AddressChangeBot', () => {
  let app: cdk.App;
  let stack: cdk.Stack;
  let template: Template;

  beforeEach(() => {
    app = new cdk.App();
    stack = new cdk.Stack(app, 'TestStack');
    new AddressChangeBot(stack, 'AddressChange', {
      name: 'test-address-change-bot',
    });
    template = Template.fromStack(stack);
  });

  test('creates a Lex bot with correct name', () => {
    template.hasResourceProperties('AWS::Lex::Bot', {
      Name: 'test-address-change-bot',
    });
  });

  test('creates a Lambda function for fulfillment', () => {
    template.hasResourceProperties('AWS::Lambda::Function', {
      Handler: 'index.handler',
    });
  });

  test('creates IAM role for Lambda execution', () => {
    template.hasResourceProperties('AWS::IAM::Role', {
      AssumeRolePolicyDocument: {
        Statement: [
          {
            Action: 'sts:AssumeRole',
            Effect: 'Allow',
            Principal: {
              Service: 'lambda.amazonaws.com',
            },
          },
        ],
      },
    });
  });

  test('creates IAM role for Lex service', () => {
    template.hasResourceProperties('AWS::IAM::Role', {
      AssumeRolePolicyDocument: {
        Statement: [
          {
            Action: 'sts:AssumeRole',
            Effect: 'Allow',
            Principal: {
              Service: 'lexv2.amazonaws.com',
            },
          },
        ],
      },
    });
  });

  test('creates AddressChange intent with required slots', () => {
    template.hasResourceProperties('AWS::Lex::Bot', {
      BotLocales: Match.arrayWith([
        Match.objectLike({
          LocaleId: 'en_US',
          Intents: Match.arrayWith([
            Match.objectLike({
              Name: 'AddressChange',
              SampleUtterances: Match.arrayWith([
                { Utterance: 'Update address' },
                { Utterance: 'Change address' },
                { Utterance: 'Address change' },
              ]),
            }),
          ]),
        }),
      ]),
    });
  });

  test('creates Agent intent for human escalation', () => {
    template.hasResourceProperties('AWS::Lex::Bot', {
      BotLocales: Match.arrayWith([
        Match.objectLike({
          LocaleId: 'en_US',
          Intents: Match.arrayWith([
            Match.objectLike({
              Name: 'Agent',
              SampleUtterances: Match.arrayWith([
                { Utterance: 'Speak to an agent' },
                { Utterance: 'Talk to a human' },
                { Utterance: 'I need human help' },
              ]),
            }),
          ]),
        }),
      ]),
    });
  });

  test('configures bot with English locale and Joanna voice', () => {
    template.hasResourceProperties('AWS::Lex::Bot', {
      BotLocales: Match.arrayWith([
        Match.objectLike({
          LocaleId: 'en_US',
          VoiceSettings: {
            Engine: 'neural',
            VoiceId: 'Joanna',
          },
        }),
      ]),
    });
  });

  test('creates Lambda permission for Lex to invoke function', () => {
    template.hasResourceProperties('AWS::Lambda::Permission', {
      Action: 'lambda:InvokeFunction',
      Principal: 'lexv2.amazonaws.com',
    });
  });

  test('creates bot version and alias', () => {
    template.hasResourceProperties('AWS::Lex::BotVersion', {
      BotVersionLocaleSpecification: [
        {
          LocaleId: 'en_US',
          BotVersionLocaleDetails: {
            SourceBotVersion: 'DRAFT',
          },
        },
      ],
    });

    template.hasResourceProperties('AWS::Lex::BotAlias', {
      BotAliasName: 'live',
    });
  });

  test('creates correct resource counts', () => {
    template.resourceCountIs('AWS::Lex::Bot', 1);
    template.resourceCountIs('AWS::Lex::BotVersion', 1);
    template.resourceCountIs('AWS::Lex::BotAlias', 1);
    template.resourceCountIs('AWS::Lambda::Function', 1);
    template.resourceCountIs('AWS::Lambda::Permission', 1);
    // 2 IAM roles: one for Lambda, one for Lex
    template.resourceCountIs('AWS::IAM::Role', 2);
  });

  test('matches snapshot', () => {
    expect(template.toJSON()).toMatchSnapshot();
  });
});
