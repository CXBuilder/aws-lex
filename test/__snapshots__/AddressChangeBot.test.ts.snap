// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddressChangeBot matches snapshot 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "AddressChangeBotAlias4B603C78": {
      "Properties": {
        "BotAliasLocaleSettings": [
          {
            "BotAliasLocaleSetting": {
              "CodeHookSpecification": {
                "LambdaCodeHook": {
                  "CodeHookInterfaceVersion": "1.0",
                  "LambdaArn": {
                    "Fn::GetAtt": [
                      "AddressChangeLambda6522C1F0",
                      "Arn",
                    ],
                  },
                },
              },
              "Enabled": true,
            },
            "LocaleId": "en_US",
          },
        ],
        "BotAliasName": "live",
        "BotId": {
          "Fn::GetAtt": [
            "AddressChangeBotE5BEDD7F",
            "Id",
          ],
        },
        "BotVersion": {
          "Fn::GetAtt": [
            "AddressChangeBotVersion9850105591AF08952",
            "BotVersion",
          ],
        },
      },
      "Type": "AWS::Lex::BotAlias",
    },
    "AddressChangeBotE5BEDD7F": {
      "Properties": {
        "AutoBuildBotLocales": false,
        "BotLocales": [
          {
            "Intents": [
              {
                "DialogCodeHook": {
                  "Enabled": false,
                },
                "FulfillmentCodeHook": {
                  "Enabled": true,
                },
                "Name": "AddressChange",
                "SampleUtterances": [
                  {
                    "Utterance": "I would like to update my address to {houseNumber} {streetName}",
                  },
                  {
                    "Utterance": "I would like to update my address to {houseNumber} {streetName} {city} {state} {zipCode}",
                  },
                  {
                    "Utterance": "I would like to update to my new address",
                  },
                  {
                    "Utterance": "I would like to update address",
                  },
                  {
                    "Utterance": "Update address",
                  },
                  {
                    "Utterance": "I want to change my address",
                  },
                  {
                    "Utterance": "Change address",
                  },
                  {
                    "Utterance": "Address change",
                  },
                ],
                "SlotPriorities": [
                  {
                    "Priority": 1,
                    "SlotName": "houseNumber",
                  },
                  {
                    "Priority": 2,
                    "SlotName": "streetName",
                  },
                  {
                    "Priority": 3,
                    "SlotName": "city",
                  },
                  {
                    "Priority": 4,
                    "SlotName": "state",
                  },
                  {
                    "Priority": 5,
                    "SlotName": "zipCode",
                  },
                ],
                "Slots": [
                  {
                    "Name": "houseNumber",
                    "SlotTypeName": "AMAZON.Number",
                    "ValueElicitationSetting": {
                      "PromptSpecification": {
                        "AllowInterrupt": true,
                        "MaxRetries": 3,
                        "MessageGroupsList": [
                          {
                            "Message": {
                              "PlainTextMessage": {
                                "Value": "What is the house or building?",
                              },
                            },
                          },
                        ],
                      },
                      "SlotConstraint": "Required",
                    },
                  },
                  {
                    "Name": "streetName",
                    "SlotTypeName": "AMAZON.AlphaNumeric",
                    "ValueElicitationSetting": {
                      "PromptSpecification": {
                        "AllowInterrupt": true,
                        "MaxRetries": 3,
                        "MessageGroupsList": [
                          {
                            "Message": {
                              "PlainTextMessage": {
                                "Value": "What is the street name?",
                              },
                            },
                          },
                        ],
                      },
                      "SlotConstraint": "Required",
                    },
                  },
                  {
                    "Name": "city",
                    "SlotTypeName": "AMAZON.City",
                    "ValueElicitationSetting": {
                      "PromptSpecification": {
                        "AllowInterrupt": true,
                        "MaxRetries": 3,
                        "MessageGroupsList": [
                          {
                            "Message": {
                              "PlainTextMessage": {
                                "Value": "What is the city?",
                              },
                            },
                          },
                        ],
                      },
                      "SlotConstraint": "Required",
                    },
                  },
                  {
                    "Name": "state",
                    "SlotTypeName": "AMAZON.State",
                    "ValueElicitationSetting": {
                      "PromptSpecification": {
                        "AllowInterrupt": true,
                        "MaxRetries": 3,
                        "MessageGroupsList": [
                          {
                            "Message": {
                              "PlainTextMessage": {
                                "Value": "What is the state?",
                              },
                            },
                          },
                        ],
                      },
                      "SlotConstraint": "Required",
                    },
                  },
                  {
                    "Name": "zipCode",
                    "SlotTypeName": "AMAZON.Number",
                    "ValueElicitationSetting": {
                      "PromptSpecification": {
                        "AllowInterrupt": true,
                        "MaxRetries": 3,
                        "MessageGroupsList": [
                          {
                            "Message": {
                              "PlainTextMessage": {
                                "Value": "What is the zip code?",
                              },
                            },
                          },
                        ],
                      },
                      "SlotConstraint": "Required",
                    },
                  },
                ],
              },
              {
                "DialogCodeHook": {
                  "Enabled": false,
                },
                "FulfillmentCodeHook": {
                  "Enabled": true,
                },
                "Name": "Agent",
                "SampleUtterances": [
                  {
                    "Utterance": "Speak to an agent",
                  },
                  {
                    "Utterance": "Talk to a human",
                  },
                  {
                    "Utterance": "I need human help",
                  },
                ],
              },
              {
                "DialogCodeHook": {
                  "Enabled": false,
                },
                "FulfillmentCodeHook": {
                  "Enabled": true,
                },
                "Name": "FallbackIntent",
                "ParentIntentSignature": "AMAZON.FallbackIntent",
              },
            ],
            "LocaleId": "en_US",
            "NluConfidenceThreshold": 0.4,
            "SlotTypes": [],
            "VoiceSettings": {
              "Engine": "neural",
              "VoiceId": "Joanna",
            },
          },
        ],
        "DataPrivacy": {
          "ChildDirected": false,
        },
        "IdleSessionTTLInSeconds": 300,
        "Name": "test-address-change-bot",
        "RoleArn": {
          "Fn::GetAtt": [
            "AddressChangeBotRole5F576B97",
            "Arn",
          ],
        },
        "TestBotAliasSettings": {
          "BotAliasLocaleSettings": [
            {
              "BotAliasLocaleSetting": {
                "CodeHookSpecification": {
                  "LambdaCodeHook": {
                    "CodeHookInterfaceVersion": "1.0",
                    "LambdaArn": {
                      "Fn::GetAtt": [
                        "AddressChangeLambda6522C1F0",
                        "Arn",
                      ],
                    },
                  },
                },
                "Enabled": true,
              },
              "LocaleId": "en_US",
            },
          ],
        },
      },
      "Type": "AWS::Lex::Bot",
    },
    "AddressChangeBotRole5F576B97": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lexv2.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "AddressChangeBotRoleLexPolicy29DD1F00": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "polly:SynthesizeSpeech",
              "Effect": "Allow",
              "Resource": "*",
            },
            {
              "Action": [
                "logs:CreateLogStream",
                "logs:PutLogEvents",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:logs:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":log-group:*:log-stream:*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "AddressChangeBotRoleLexPolicy29DD1F00",
        "Roles": [
          {
            "Ref": "AddressChangeBotRole5F576B97",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "AddressChangeBotVersion9850105591AF08952": {
      "Properties": {
        "BotId": {
          "Fn::GetAtt": [
            "AddressChangeBotE5BEDD7F",
            "Id",
          ],
        },
        "BotVersionLocaleSpecification": [
          {
            "BotVersionLocaleDetails": {
              "SourceBotVersion": "DRAFT",
            },
            "LocaleId": "en_US",
          },
        ],
      },
      "Type": "AWS::Lex::BotVersion",
    },
    "AddressChangeLambda6522C1F0": {
      "DependsOn": [
        "AddressChangeLambdaServiceRoleA0F935CC",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "925178554ba0ba8388de51d6c7d0154c5d56a4a5be5d9abca7dd6ffcae7b6a20.zip",
        },
        "Environment": {
          "Variables": {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
          },
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "AddressChangeLambdaServiceRoleA0F935CC",
            "Arn",
          ],
        },
        "Runtime": "nodejs16.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "AddressChangeLambdaServiceRoleA0F935CC": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "AddressChangeLambdalexlambdainvokeenUS115EC9B5": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "AddressChangeLambda6522C1F0",
            "Arn",
          ],
        },
        "Principal": "lexv2.amazonaws.com",
        "SourceArn": {
          "Fn::Join": [
            "",
            [
              "arn:aws:lex:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":bot-alias/",
              {
                "Ref": "AddressChangeBotE5BEDD7F",
              },
              "/*",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
